### PR TITLE
sub-PR 5

### PR DESCRIPTION
**Pull request by drichte2**
Hopefully there aren't too many things in this one. Several different miscellaneous items, including some new statistics.

### CODE REVIEW

File: src/parcel.F
Comment (sjsprecious) [inline src/parcel.F:1224]: why removing this line?

File: src/param.F
Comment (sjsprecious) [inline src/param.F:6087]: why adding `npvars` by 1 here?

File: src/parcel.F
Comment (drichte2) [reply src/parcel.F:1224]: I wanted each particle to know how long it have lived (i.e. stored in pdata). But then I realized that there was already a pdata "prtime" that wasn't being used, so I just repurposed it. It used to simply record the model time, but now I changed it so that it provides total accumulated time since the particle was created.

File: src/param.F
Comment (drichte2) [reply src/param.F:6087]: See above, but it would need to be stored at restart. It's not a diagnostic (like the interpolated quantities are)

File: src/droplet.F
Comment (sjsprecious) [inline src/droplet.F:24]: @drichte2 Do these variables have to be module variables? They are used in the `acc routine` and involved in the reduction operation, which makes the GPU implementation difficult. Would like to see @johnmauff's option, too.

File: src/param.F
Comment (sjsprecious) [reply src/param.F:6087]: thanks for the clarification.

File: src/parcel.F
Comment (sjsprecious) [reply src/parcel.F:1224]: thanks for the clarification.

File: src/droplet.F
Comment (drichte2) [reply src/droplet.F:24]: Technically, there would be a way to remove these as module variables, but it'd be a pain. They'd have to be sent out of the droplet_driver and then into droplet_diag, which would mean declaring them somewhere at the cm1.F level. I suppose this could be done, but this seemed like a more compact way of doing it. 

For what it's worth, @johnmauff said in an email when I asked about this, "I don't think this should cause a problem for the GPU port, however, I want to hear Jian's opinion as well.   We can just mark the new data structures as device resident and make sure that the correct version gets updated."

File: src/droplet.F
Comment (sjsprecious) [inline src/droplet.F]: shall we remove these three variables here since they are module variables now?

Comment (drichte2) [reply src/droplet.F]: Yes, sorry -- that got missed when fixing the merge conflicts. Note that num0 and num1 need to stay where they are. Those that got moved to the module should be removed (in fact, I don't understand how this compiled for me...)

Comment (sjsprecious) [reply src/droplet.F]: Thanks @drichte2 for the clarification and fix.

File: src/droplet.F
Comment (sjsprecious) [reply src/droplet.F:24]: Generally I think John's comment is correct. However, in this particular case, those module variables are involved in a reduction operation in an acc routine, and I feel it is not trivial to get the code right on the GPU. I have added the necessary GPU directives (I think) and work on the testing now.

Comment (johnmauff) [reply src/droplet.F:24]: I am not particularly supportive of adding additional arguments that are passed through the subroutine argument list.  However, Jian is correct that the use of module variables complicates the GPU implementation.  Currently, CM1 has a qbudget array that stores a number of different global sums related to moisture.  Maybe something similar could be done with droplet statistics if Jian's testing on GPU proves problematic.

File: src/droplet.F
Comment (sjsprecious) [inline src/droplet.F:2127]: @drichte2 Shall we remove line 2128 to 2139 here? I saw similar calculations in the `initialize_droplet_histograms` subroutine and they did not seem to vary with different time steps.

Comment (drichte2) [reply src/droplet.F:2127]: This is strange... Yes, you are absolutely right. But somehow these lines aren't in my local version that I had pushed. I must have messed up something when I was merging? I'd change this myself, but I don't see how. Could have this come back in when you merged gpu-opt back into this branch?

Comment (sjsprecious) [reply src/droplet.F:2127]: Thanks @drichte2. I did not know why this happened, either. I just removed these lines and pushed the changes to your branch. Could you pull the latest version of your branch to your local machine and check if everything looks good to you now?

Comment (drichte2) [reply src/droplet.F:2127]: Just pulled, looks good to me!

Comment (sjsprecious) [reply src/droplet.F:2127]: Thanks!

**Comment by sjsprecious** [comment]
By using the namelist `namelist.restC` and without the conda environment, the verification result on Casper looks below:

---

CPU (intel, 2 MPI ranks) vs. GPU (nvhpc, 2 MPI ranks + 2 V100 GPUs)

Stat output comparison

Identical variables: 63
variables with rel_diff > 1.e-6: 23
variables with rel_diff <= 1.e-6: 0
four variables with largest error: PPMIN, PPMAX, VOR2KM, DIVMAX

Droplet 0th-order diagnostic comparison

Identical variables: 9
variable with rel_diff > 1.e-6: 3
variable with rel_diff <= 1.e-6: 0
three variables with largest error: Tpavg (~3.4e-6), radmin (~6.2e-6), prsavg (~1.1e-6)

Droplet 1st-order diagnostic comparison

Identical variables: 511
variable with rel_diff > 1.e-6: 1
variable with rel_diff <= 1.e-6: 0
one variables with largest error: numconc(1) (~1.0)

---

Compared to the PR #109, the stat output comparison result is consistent. The 0th-order droplet diagnostic shows one fewer identical variable and one more variable with relative difference >1.e-6, which is `prsavg`. The 1st-order droplet diagnostic shows one inconsistent variable, which is `numconc(1)`. The relative difference is close to 100%. The value of `numconc(1)` in the CPU run is **0.150681E-04** while that in the GPU run is **0.152588E-07**. I am not sure whether this difference is caused by the mismatch of `prsavg` variable or not. I am also not sure whether the difference of the droplet diagnostic here is even expected or not, though the `init3d` subroutine is tweaked slightly here. Need to see @drichte2 and @johnmauff's opinions on these results.

**Comment by drichte2** [comment]
@sjsprecious I'm pretty sure the answer is yes, but `tnumpart` is part of the 0th order diagnostics, right? And that comes out exactly the same (it'd better)? The difference in `prsavg` should have nothing to do with the 1st order diagnostic. 

It would be very weird if `tnumpart` was the same but only one element of `numconc` was off, since the sum of the `numconc` array should be equal to `tnumpart` divided by the domain volume.

**Comment by drichte2** [comment]
@sjsprecious For this issue, maybe instead of checking `numconc` you can replace it with `partnum`? This variable simply says how many particles are within each height range. The sum of this array should give you exactly `tnumpart`, and it's `partnum` which is subject to reduction operations (and therefore where the difference might come from)

**Comment by sjsprecious** [comment]
> @sjsprecious I'm pretty sure the answer is yes, but `tnumpart` is part of the 0th order diagnostics, right? And that comes out exactly the same (it'd better)? The difference in `prsavg` should have nothing to do with the 1st order diagnostic.
> 
> It would be very weird if `tnumpart` was the same but only one element of `numconc` was off, since the sum of the `numconc` array should be equal to `tnumpart` divided by the domain volume.

Hi @drichte2, thanks for your quick reply. Yes, `tnumpart` matches exactly between CPU and GPU runs, which makes me think if something is wrong when calculating the domain volume. I originally thought `prsavg` was the culprit but it seems not now.

**Comment by sjsprecious** [comment]
> @sjsprecious For this issue, maybe instead of checking `numconc` you can replace it with `partnum`? This variable simply says how many particles are within each height range. The sum of this array should give you exactly `tnumpart`, and it's `partnum` which is subject to reduction operations (and therefore where the difference might come from)

Hi @drichte2, thanks for your suggestion. I just realized that I commented out some GPU directives in the `droplet_diag` subroutine and therefore the 1st-order droplet diagnostic was problematic. Somehow I could not get those directives work for the changes in this PR and I would need more time to debug them. Sorry for the confusion.

**Comment by sjsprecious** [comment]
Hmm, it turns out that the GPU code with the changes here no longer compiles on Casper (nvhpc/22.2 + cuda/11.6), but still compiles on Gust (nvhpc/23.1 + cuda/11.7). Thus maybe our GPU changes here trigger a bug in the old nvhpc compiler.

By using the namelist `namelist.restC` and without the conda environment, the verification result on **Gust** looks below:

---

CPU (intel, 2 MPI ranks) vs. GPU (nvhpc, 2 MPI ranks + 2 A100 GPUs)

Stat output comparison

Identical variables: 59
variables with rel_diff > 1.e-6: 22
variables with rel_diff <= 1.e-6: 5
four variables with largest error: PPMIN, PPMAX, VOR2KM, DIVMAX

Droplet 0th-order diagnostic comparison

Identical variables: 8
variable with rel_diff > 1.e-6: 3
variable with rel_diff <= 1.e-6: 1
three variables with largest error: Tpavg (~3.4e-6), radmin (~6.2e-6), prsavg (~1.1e-6)

Droplet 1st-order diagnostic comparison

Identical variables: 512
variable with rel_diff > 1.e-6: 0
variable with rel_diff <= 1.e-6: 0

---

Compared to the PR https://github.com/george-bryan/CM1/pull/109, the stat output comparison result shows four fewer identical variable, one fewer variable with relative difference >1.e-6, and five more variables with relative difference <1.e-6. These five variables are `THPMAX`, `STHPMX`, `TKEMAX` , `RHMIN` and `THPMIN`. The 0th-order droplet diagnostic shows two fewer identical variables, one more variable with relative difference >1.e-6 (which is `prsavg`), and one more variable with relative difference <1.e-6 (which is `radavg`). The 1st-order droplet diagnostic is consistent. Since I switch to a different machine, I expect that the verification results may change and most non-identical variables are still within the tolerance. Still need to see @drichte2 and @johnmauff's comments on these results.

**Comment by drichte2** [comment]
> droplet diagnostic shows two fewer identical variables, one more variable with relative difference >1.e-6 (which is `prsavg`), and one more variable with relative difference <1.e-6 (which is `radavg`). The 1st-order droplet diagnostic is consistent. Since I switch to a different machine, I expect that the verification results may change and most non-identical variables are still within the tolerance. Still need to see @drichte2 and @johnmauff's comments on these results.

@sjsprecious Just to be clear, these are relative differences, right? For example the droplet radii are O(1e-6) so an absolute difference that large is like 100% relative error. 

Also, just to confirm: these are a comparison between CPU and GPU of this most recent PR -- not a comparison of the previous version to the PR?

Assuming these are true, none of those numbers are particularly concerning to me. It seems that those that are larger than 1e-6 are still O(1e-6).

**Comment by sjsprecious** [comment]
> > droplet diagnostic shows two fewer identical variables, one more variable with relative difference >1.e-6 (which is `prsavg`), and one more variable with relative difference <1.e-6 (which is `radavg`). The 1st-order droplet diagnostic is consistent. Since I switch to a different machine, I expect that the verification results may change and most non-identical variables are still within the tolerance. Still need to see @drichte2 and @johnmauff's comments on these results.
> 
> @sjsprecious Just to be clear, these are relative differences, right? For example the droplet radii are O(1e-6) so an absolute difference that large is like 100% relative error.
> 
> Also, just to confirm: these are a comparison between CPU and GPU of this most recent PR -- not a comparison of the previous version to the PR?
> 
> Assuming these are true, none of those numbers are particularly concerning to me. It seems that those that are larger than 1e-6 are still O(1e-6).

@drichte2 , yes, all the difference of statistics here is relative difference.

And yes, the comparison here is between CPU and GPU runs for this PR.

Glad to know that these numbers look fine to you. Will wait for John's input before I approve this PR.

**Review by sjsprecious — APPROVED** [review]

**Review by johnmauff — APPROVED** [review]
Based on Jian extensive review and testing, and discussion with him, I approve this PR.

**Pull request merged**
